com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/AbstractProtocolServer.java

```diff
@@ -102,14 +117,80 @@ protected void sendMessage(Messages.ProtocolMessage message) {
         String utf8Data = data.getString(PROTOCOL_ENCODING);
 
         try {
-            logger.fine("\n[[RESPONSE]]\n" + utf8Data);
+            if (message instanceof Messages.Request) {
+                logger.fine("\n[[REQUEST]]\n" + utf8Data);
+            } else if (message instanceof Messages.Event) {
+                logger.fine("\n[[EVENT]]\n" + utf8Data);
+            } else {
+                logger.fine("\n[[RESPONSE]]\n" + utf8Data);
+            }
             this.writer.write(utf8Data);
             this.writer.flush();
         } catch (IOException e) {
             logger.log(Level.SEVERE, String.format("Write data to io exception: %s", e.toString()), e);
         }
     }
 
+    @Override
+    public void sendEvent(DebugEvent event) {
+        sendMessage(new Messages.Event(event.type, event));
+    }
+
+    @Override
+    public void sendResponse(Messages.Response response) {
+        sendMessage(response);
+    }
+
+    @Override
+    public void sendRequest(Messages.Request request, Consumer<Messages.Response> cb) {
+        sendRequest(request, 0, cb);
+    }
+
+    @Override
+    public void sendRequest(Messages.Request request, long timeout, Consumer<Messages.Response> cb) {
+        Timer timer = new Timer();
+        Disposable[] disposable = new Disposable[1];
+        boolean[] executed = new boolean[]{false};
+        disposable[0] = subject.filter(response -> response.request_seq == request.seq).subscribe((response) -> {
+            synchronized (executed) {
+                if (executed[0]) {
+                    return;
+                } else {
+                    executed[0] = true;
+                    timer.cancel();
+                }
+            }
+            cb.accept(response);
+            if (disposable[0] != null) {
+                disposable[0].dispose();
+                disposable[0] = null;
+            }
+        });
+        sendMessage(request);
+        if (timeout > 0) {
+            try {
+                timer.schedule(new TimerTask() {
+                    @Override
+                    public void run() {
+                        synchronized (executed) {
+                            if (executed[0]) {
+                                return;
+                            }
+                            executed[0] = true;
+                        }
+                        if (disposable[0] != null) {
+                            disposable[0].dispose();
+                            disposable[0] = null;
+                        }
+                        cb.accept(new Messages.Response(request.seq, request.command, false, "timeout"));
```

Comment:
Reporting timeout error using the normal execution flow is not recommended. Please switch to CompletableFuture and try to completeExceptionally with a TimeoutException.